File: Sorting-Algorithms/Counting_Sort/Java/counting_sort.java
/*
    Counting Sort
    Time Complexity
        Best Case Scenario  : O(n+k)
        Average Case Scenario  : O(n+k)
        Worst Case Scenario : O(n+k)
*/
import java.util.*;
public class counting_sort
{
    static public void countingSort(int A[],int n)
    {
        int k=A[0],i;
        int output[]=new int[n];
        for(i=0;i<n;i++)
        {
            if(k<A[i])
            {
                k=A[i];
            }
        }
        int count[]=new int[k+1];
        for(i=0;i<=k;i++)
        {
            count[i]=0;
        }
        for(i=0;i<n;i++)
        {
            count[A[i]]++;
        }
        for(i=1;i<=k;i++)
        {
            count[i]+=count[i-1];
        }
        for(i=n-1;i>=0;i--)
        {
            output[--count[A[i]]]=A[i];
        }
        for(i=0;i<n;i++)
        {
            A[i]=output[i];
        }
    }
    public static void main(String args[])
    {
        int arr[]={1,10,30,20,2,11,15};
        int n=arr.length;
        countingSort(arr,n);
        System.out.print("Array after counting sort : ");
        for(int i=0;i<n;i++)
        {
            System.out.print(arr[i]+" ");
        }
        System.out.println();
    }
}

File: Sorting-Algorithms/Counting_Sort/Python/counting_sort.py
'''
    Counting Sort
    Time Complexity
        Best Case Scenario  : O(n+k)
        Average Case Scenario  : O(n+k)
        Worst Case Scenario : O(n+k)
'''
def countingSort(A,n):
    k=A[0]
    output=[0]*(n)
    for i in range(0,n):
        if(k<A[i]):
            k=A[i]
    
    count=[0]*(k+1)
    for i in range(0,n):
        count[A[i]]+=1
    for i in range(1,k+1):
        count[i]+=count[i-1]
    for i in range(n-1,-1,-1):
        count[A[i]]-=1
        output[count[A[i]]]=A[i]
    for i in range(0,n):
        A[i]=output[i]

arr=[1,10,30,20,2,11,15]
n=len(arr)
countingSort(arr,n)
print("Array after counting sort : ", end='')
for i in arr:
    print("{} ".format(i),end="")

print()
